spec.bs

```diff
@@ -1902,6 +1936,65 @@ To <dfn>convert a modifier to a string</dfn> given a [=part/modifier=] |modifier
+
+1. **Accept shorthands**. Most author patterns will be simple and straightforward. Accordingly, APIs should accept shorthands for those common cases and avoid the need for authors to take additional steps to transform these into complete {{URLPattern}} objects.
+1. **Respect the base URL**. Just as URLs are generally parsed relative to a base URL for their environment (most commonly, a [=document base URL=]), URL patterns should respect this as well. The {{URLPattern}} constructor itself is an exception because it directly exposes the concept itself, similar to how the {{URL}} constructor does not respect the base URL even though the rest of the platform does.
+1. **Be clear about regexp groups**. Some APIs may benefit from only allowing URL patterns which do not [=URLPattern/have regexp groups=], for example, because user agents are likely to implement them in a different thread or process from those executing author script, and because of security or performance concerns, a JavaScript engine would not ordinarily run there. If so, this should be clearly documented (with reference to {{URLPattern/hasRegExpGroups}}) and the operation should report an error as soon as possible (e.g., by throwing a JavaScript exception). If possible, this should be feature-detectable to allow for the possibility of this constraint being lifted in the future. Avoid creating different subsets of URL patterns without consulting the editors of this specification.
+1. **Be clear about what URLs will be matched**. For instance, algorithms during fetching are likely to operate on URLs with no [=url/fragment=]. If so, the specification should be clear that this is the case, and may advise showing a developer warning if a pattern which cannot match (e.g., because it requires a non-empty fragment) is used.
+
+<h3 id=integrating-javascript>Integrating with JavaScript APIs</h3>
+
+JavaScript APIs should accept all of:
+* a {{URLPattern}} object
+* a dictionary-like object which specifies the components required to construct a pattern
+* a string (in the constructor string syntax)
+
+To accomplish this, specifications should accept {{URLPatternCompatible}} as an argument to an [=operation=] or [=dictionary member=], and process it using the following algorithm, using the appropriate [=environment settings object=]'s [=environment settings object/API base URL=] or equivalent.
+
+<div algorithm>
+  To <dfn for=URLPattern>build a {{URLPattern}} from a WebIDL value</dfn> {{URLPatternCompatible}} |input| given [=URL=] |baseURL|, perform the following steps:
+
+  1. If the [=specific type=] of |input| is {{URLPattern}}:
+    1. Return |input|.
+  1. Otherwise, if the [=specific type=] of |input| is {{URLPatternInit}}:
+    1. If |input|["`baseURL`"] does not [=map/exist=], set it to the [=URL serializer|serialization=] of |baseURL|.
+    1. Return the result of constructing a {{URLPattern/constructor(input, options)}} given |input|.
```

Comment:
This isn't super well-defined, and it'd be more well-founded if we factored out the constructor steps into separate algorithms and called them. (Maybe after doing a [new](https://webidl.spec.whatwg.org/#new) `URLPattern`.) This ties into a general issue where we prefer other specs not to operate directly on the public API, but that's a larger refactoring.

If we don't want to do that kind of work, then I think the best way to do this would be something like

> 1. Let _urlPattern_ be a [new](https://webidl.spec.whatwg.org/#new) `URLPattern`.
> 2. Perform the `new URLPattern(input, baseURL, options)` [constructor steps](https://webidl.spec.whatwg.org/#constructor-steps) with _urlPattern_ as [this](https://webidl.spec.whatwg.org/#this) and « _input_, the serialization of _baseURL_ » as argument values.

Reply:
Hmm, since [initialize](https://whatpr.org/urlpattern/199.html#urlpattern-initialize) already exists, it might be better to just call that directly.

Reply:
Done.